kyuubi-server/src/main/scala/org/apache/kyuubi/engine/EngineRef.scala

```diff
@@ -56,7 +58,21 @@ private[kyuubi] class EngineRef private(conf: KyuubiConf, user: String, sessionI
   // Share level of the engine
   private val shareLevel: ShareLevel = ShareLevel.withName(conf.get(ENGINE_SHARE_LEVEL))
 
+  // This a is a server-side option
+  val poolThreshold: Int = KyuubiConf().loadFileDefaults().get(ENGINE_POOL_SIZE_THRESHOLD)
+
   private val subDomain: Option[String] = conf.get(ENGINE_SHARE_LEVEL_SUB_DOMAIN)
+    .orElse {
+      Option(conf.get(ENGINE_POOL_SIZE))
+        .filter(size => size > 0)
+        .map(size => {
+          val poolSize = size.min(poolThreshold)
+
+          // TODO: Currently, we use random policy,
+          //  and later we can add a sequential policy, such as AtomicInteger % poolSize.
+          "engine-pool-" + Random.nextInt(poolSize)
```

Comment:
ENGINE_SHARE_LEVEL_SUB_DOMAIN has restrictions for the values,`^[a-zA-Z_]{1,10}$`, we need improve that as the prefix `engine-pool-` breaks this rule

Reply:
Got